common/xPowerManager.h

```diff
@@ -0,0 +1,769 @@
+            for (uint8_t i = 0; i < 16; i++)
+            {
+                if((pinMap & (1 << i)) )
+                {
+                    stm32l4_gpio_pin_configure(pinNRoffset + i, GPIO_MODE_ANALOG | GPIO_OTYPE_PUSHPULL | GPIO_OSPEED_LOW | GPIO_PUPD_NONE);
+                }    
+            }
+
+#ifdef DIAGNOSE_POWER_PORT
+            STDOUT.print("PORT ");STDOUT.println(targetPortAddr);
+            printGPIOState(pPortX);
+#endif
+
+        }
+
+        static void  CPU_DeepSleep(uint8_t STOPEntry) // __attribute__((optimize("O0"))) 
+        {
+            GPIO_TypeDef portA, portB, portC, portH;
+            wakeUpSource = wakeUp_none;            // reset wake up source
+
+            armv7m_systick_disable();               // disable systick 
+            
+            savePORTState(&portA, GPIOA_BASE);
+            savePORTState(&portB, GPIOB_BASE);
+            savePORTState(&portC, GPIOC_BASE);
+            savePORTState(&portH, GPIOH_BASE);
+            
+            // put all pins of port A to default, except :
+            defaultPortState(GPIOA_BASE, 0b1111111111111111);
+            // put all pins of port B to default , except: 
+            // PB2 - VUSB , PB6- BTN1 , PB14 - SD ENABLE
+            defaultPortState(GPIOB_BASE, 0b1011111110111011);
+
+            defaultPortState(GPIOC_BASE, 0b1110000000000000);
+            
+            defaultPortState(GPIOH_BASE, 0b0000000000001000);  
+
+            initWakeups();
+            // MODIFY_REG(PWR->CR1, PWR_CR1_LPMS, PWR_CR1_LPMS_STOP0); // Stop 0 mode with Main Regulator
+            MODIFY_REG(PWR->CR1, PWR_CR1_LPMS, PWR_CR1_LPMS_STOP1); // Stop 1 mode with Main Regulator OFF
+            SET_BIT(SCB->SCR, ((uint32_t)SCB_SCR_SLEEPDEEP_Msk));   // Set SLEEPDEEP bit of Cortex System Control Register
+
+            if(STOPEntry == PWR_STOPENTRY_WFI)      // Select Stop mode entry 
+                __WFI();    // Request Wait For Interrupt
+            else {
+                __SEV();    // Request Wait For Event
+                __WFE();
+                __WFE();
+            }
+            // WAKE up haved occured
+            CLEAR_BIT(SCB->SCR, ((uint32_t)SCB_SCR_SLEEPDEEP_Msk)); // Reset SLEEPDEEP bit of Cortex System Control Register
+            deinitWakeups();
+
+            stm32l4_system_initialize(_SYSTEM_CORE_CLOCK_, _SYSTEM_CORE_CLOCK_/2, _SYSTEM_CORE_CLOCK_/2, STM32L4_CONFIG_LSECLK, STM32L4_CONFIG_HSECLK, STM32L4_CONFIG_SYSOPT);
+
+            restorePORTXState(&portA, GPIOA_BASE);
+            restorePORTXState(&portB, GPIOB_BASE);
+            restorePORTXState(&portC, GPIOC_BASE);
+            restorePORTXState(&portH, GPIOH_BASE);
+
+            armv7m_systick_enable(); 
+        }
+
+    } powerman;     // global object!
+
+    // POWER DOMAINS
+    // ==================================================================================
+    // Power domain objects must inherit from this class to get registered. 
+    class xPowerDomain {
```

Comment:
Since all the power domains are hardcoded, it seems unnecessary to dynamically build a list of powerdomain classes. Seems like we could have an array of them, or even just hardcode them all in the power manager itself.

Alternatively, maybe there should be one power-manager per domain. Using bitfields to turn things on/off doesn't seem all that useful.

Reply:
The idea was to ease integration.  Hardcoding or even a static list would require maintenance, the linked list allows adding power domains by simply inheriting from a class. 

One power manager per domain wouldn't work, its job is to handle the overall interaction between domains and subscribers. Bitfields (or something to that effect) are needed to allow more than one-on-one connections between domains and subscribers - for instance the dac to subscribe to amplifier and booster, or  multiple subscribers to subscribe to the cpu domain.

Reply:
You already have a static list in the bitfield though, so it's already hardcoded anyways.
Using linked lists just takes up code space.